src/AssemblyAI/Files/FilesClient.cs

```diff
@@ -19,21 +20,59 @@ public FilesClient(RawClient client)
     /// <summary>
     /// Upload a media file to AssemblyAI's servers.
     /// </summary>
-    public async Task<UploadedFile> UploadAsync(Stream request)
+    public async Task<UploadedFile> UploadAsync(Stream request, RequestOptions? options = null)
     {
         var response = await _client.MakeRequestAsync(
             new RawClient.StreamApiRequest
             {
+                BaseUrl = _client.Options.BaseUrl,
                 Method = HttpMethod.Post,
                 Path = "v2/upload",
-                Body = request
+                Body = request,
+                Options = options
             }
         );
         var responseBody = await response.Raw.Content.ReadAsStringAsync();
         if (response.StatusCode is >= 200 and < 400)
         {
-            return JsonSerializer.Deserialize<UploadedFile>(responseBody)!;
+            try
+            {
+                return JsonUtils.Deserialize<UploadedFile>(responseBody)!;
+            }
+            catch (JsonException e)
+            {
+                throw new AssemblyAIClientException("Failed to deserialize response", e);
+            }
+        }
+
+        try
+        {
+            switch (response.StatusCode)
+            {
+                case 400:
+                    throw new BadRequestError(JsonUtils.Deserialize<Error>(responseBody));
+                case 401:
+                    throw new UnauthorizedError(JsonUtils.Deserialize<Error>(responseBody));
+                case 404:
+                    throw new NotFoundError(JsonUtils.Deserialize<Error>(responseBody));
+                case 429:
+                    throw new TooManyRequestsError(JsonUtils.Deserialize<Error>(responseBody));
+                case 500:
+                    throw new InternalServerError(JsonUtils.Deserialize<Error>(responseBody));
+                case 503:
+                    throw new ServiceUnavailableError(JsonUtils.Deserialize<object>(responseBody));
+                case 504:
+                    throw new GatewayTimeoutError(JsonUtils.Deserialize<object>(responseBody));
+            }
```

Comment:
For exceptions that have a JSON error response, we want the exception message to be set to the `.error` message.
Having the error code is helpful too (as a property) but shouldn't be included in the message.

Reply:
> we want the exception message to be set to the .error message.
We are doing this today, no? I may be misunderstanding 😄 

Do you mean that you want the `.error` message to be the `Exception` message?

Reply:
If I catch the exception, this should be true:
```csharp
Assert.That(ex.Message, Is.EqualTo("each transcript source id must be valid"));
```

Currently, the `ex.Message` is "BadRequestError".

Reply:
Hey @Swimburger , this ask makes a lot of sense, but it's a heavy lift on our end, as it would be a new end to end feature

Would it be possible to remove this from the GA blockers, and we'll follow up with it after GA?

Reply:
As I suggested in a past conversation, we don't need this to be something that Fern generates, but we do need an extension point where we can add our own code without disabling too much code generation.

Reply:
You can provide a partial method for us to implement like I show here:
https://github.com/AssemblyAI/assemblyai-csharp-sdk/pull/39/files#diff-cadbe42714ba9be588f87ec2595828c853eed16be025736993d2c370bb9e6e1aR10